### PR TITLE
Cfin 263 transform api response

### PR DESCRIPTION
**Pull request by RichardAllum**
I think there are definitely improvements that can be made to our tests, in particular a full happy path test with transformations, but I wanted to get some feedback on the current state of the API.

### CODE REVIEW

File: src/utils/transformResponse.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const transformDepartment = (course) => {
+    if(course && course.department) {
+        const departments = typeof course.department === "string"
+            ? course.department
+            : "";
+
+        if(departments.includes(", ")) {
+            course.department = departments.split(", ");
```
Comment (RichardAllum) [inline src/utils/transformResponse.js:35]: We have 2 departments: Theatre, Film, Television and Interactive media/Philosophy, Politics and Economics that have commas in their names. Is there a better exception than hard-coding these exceptions?

File: src/utils/transformResponse.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * our OpenAPI specification.
+ * @param response {[Object]}
+ * @returns {[Object]}
+ */
+module.exports.transformResponse = (response) => {
+    return response.map(course => {
+        transformDistanceLearning(course);
+        transformDepartment(course);
+        return course;
+    });
```
Comment (ce666-york) [inline src/utils/transformResponse.js:12]: Would it be better/more conventional to treat `course` as immutable?

File: src/utils/transformResponse.test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        ];
+
+        const result = transformResponse(weirdInputs);
+
+        expect(result[0].distanceLearning).toBeNull();
+        expect(result[1].distanceLearning).toBeNull();
+        expect(result[2].distanceLearning).toBeNull();
+    });
+
+    it('splits departments into an array', () => {
```
Comment (ce666-york) [inline src/utils/transformResponse.test.js:47]: I think it would be better to split the single/multiple department test cases out into separate tests

File: src/utils/transformResponse.js [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+const transformDistanceLearning = (course) => {
+    if(course && course.distanceLearning) {
+        const distanceLearning = typeof course.distanceLearning === "string"
+          ? course.distanceLearning.toLowerCase()
+          : null;
+
+        course.distanceLearning = distanceLearning === "yes" ? true : distanceLearning === "no" ? false : null;
+    }
+
+    return course;
```
[the remark this answers is not in the document]
Comment (ce666-york) [reply src/utils/transformResponse.js:25]: It might also be worth verifying that the Funnelback OpenAPI spec is correct with the current dataset

**Review by ce666-york — APPROVED** [review]
I think this is ok to be merged, I've put a few suggestions in too

File: src/utils/transformResponse.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/**
+ * Generic function to map the response from our Funnelback API so that it matches
+ * our OpenAPI specification.
+ * @param response {[Object]}
+ * @returns {[Object]}
+ */
```
Comment (FergusMcGlynn) [inline src/utils/transformResponse.js:6]: I'm not against JSDoc per se, but I'm not massively keen on it being used sometimes and not others. Also, some of the JS aficionados in ESG really don't like JSDoc while others do. My inexperienced opinion is that by default we should stick to clean code principles which I believe would advise against JSDoc generally, unless we decided as a team to adopt them.

File: src/utils/transformResponse.js (hunk 1 above)
Comment (FergusMcGlynn) [reply src/utils/transformResponse.js:35]: Do we have any control over the separator used in the Funnelback results? If so, I'd say a better solution would be to use a separator other than comma, e.g. pipe?

File: src/handler.test.js [hunk 6]
```diff
@@ -92,6 +92,45 @@ test('Response with 200 code is returned correctly', async () => {
     expect(result.body).toEqual('{"results":[]}');
 });
 
+test('Response with basic results returns those results', async () => {
+    const event = {
+        queryStringParameters: {
+            search: 'physics',
+        }
+    };
```
Comment (FergusMcGlynn) [inline src/handler.test.js:100]: I'm a little unsure about the value of this test as it stands. I'm not quite sure of its purpose. Could it be amended to include the fields that need transforming, and then it would demonstrate that "this response from funnelback leads to this response from the Courses API"? As it stands it's kind of saying "this partial response from funnelback gets passed straight through the Courses API (because it doesn't need transforming)" which doesn't feel as satisfying to me. Alternatively, rename the test to be explicit that its specifically testing responses that don't need transforming. But then, is that a particularly valuable test?

File: src/utils/transformResponse.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+
+const transformDistanceLearning = (course) => {
+    if(course && course.distanceLearning) {
+        const distanceLearning = typeof course.distanceLearning === "string"
+          ? course.distanceLearning.toLowerCase()
+          : null;
+
+        course.distanceLearning = distanceLearning === "yes" ? true : distanceLearning === "no" ? false : null;
```
Comment (FergusMcGlynn) [inline src/utils/transformResponse.js:22]: I find nested ternary expressions a little hard to follow personally - not a big deal though!

File: src/utils/transformResponse.js (hunk 1 above)
Comment (ncn1) [reply src/utils/transformResponse.js:35]: Do we know how this specific metadata item on the course pages is being created? ie. is it just free-format entry, or is there a list of predefined items? If free-format entry, there will be extra difficulties with spaces and positioning of commas when trying to hard-code. What problem are we solving by splitting this into an array at this point? If there is nothing that currently requires us do this rather than just displaying the string as supplied from Funnelback, it would seem to be violating the advance optimisation rule to split it up now.

File: src/handler.test.js (hunk 6 above)
Comment (ce666-york) [reply src/handler.test.js:100]: I think the value it offers comes from regression testing, I'd be happy with the test description reflecting that more clearly

File: src/utils/transformResponse.js (hunk 1 above)
Comment (ce666-york) [reply src/utils/transformResponse.js:35]: I don't think that agreeing on a model that accurately reflects the data (in this case, that `department` is a list and should therefore be represented as an array) is too premature, though I see your point that right now it's not affecting anything.

I agree using a safe delimiter would be better than hard-coding, but I don't know much about how the metadata is maintained right now (I think this is in the BA backlog, or at least on the radar).

Comment (ncn1) [reply src/utils/transformResponse.js:35]: Depends how we interpret 'accurately reflects the data' - at the moment, we can only be sure that our department data is a string, and to a human, it can look like a list. But  if we don't know of anything enforcing a particular format, it's only a free-format string as far as the computer is concerned.
